### PR TITLE
Chainlink Functions | Gateway

### PR DESCRIPTION
**Pull request by uklok-agent**
## Description

A unified Gateway for all the Chainlink Function interactions was created.
It will be used to abstract all the required logic from the different places we need to use this feature and simplified as follows:

- The gateway will have registered all the different codes that could be run. -- Imagine this as an on-chain `API` --
- Each "code" will have an independent subscription tied to it. -- Imagine each code/function is an endpoint and the `path` is the subscription ID--
- The code and configuration could be updated whenever it's needed, so NO new contracts/subscriptions/deployments are required for this.
- The gateway has an ACL so only allowed Managers could update endpoints, and only registered clients can use the API.
- Each client could use 1 or more "endpoints" as each response contains the "subscription Id" (endpoint) requested. This means a simple contract implementation where several endpoints are required could be easily managed by a `switch(subscriptionId)` where each case is an internal call to the corresponding handler.

> example:
```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.0;

import {ICCGatewayClient} from "./interfaces/ICCGatewayClient.sol";
import {ICCGateway} from "./interfaces/ICCGateway.sol";

contract CCExampleClient is ICCGatewayClient {
    ICCGateway private immutable gateway;
    uint64 private chat_gpt_endpoint = 234;
    uint64 private star_wars_endpoint = 235;

    constructor(address gatewayAddress) {
        gateway = ICCGateway(gatewayAddress);
    }

    function requestGPT(string[] calldata args, bytes[] calldata bytesArgs) external {
        gateway.sendRequest(chat_gpt_endpoint, args, bytesArgs, "");
    }

    function handleGPTResponse(ICCGatewayClient.CCGResponse calldata response)  {
        if (response.state == ICCGatewayClient.CCGResponseState.Success) {
            // Do something with the `response.data`
        } else {
            // Handle the `response.error`
        }
    }

    function requestStarWars(string[] calldata args, bytes[] calldata bytesArgs) external {
        gateway.sendRequest(star_wars_endpoint, args, bytesArgs, "");
    }

    function handleStarWarsResponse(ICCGatewayClient.CCGResponse calldata response)  {
        if (response.state == ICCGatewayClient.CCGResponseState.Success) {
            // Do something with the `response.data`
        } else {
            // Handle the `response.error`
        }
    }

    function callback(bytes32 requestId) external override {
        ICCGatewayClient.CCGResponse memory response = gateway.getResponse(requestId, true);

        if (response.subscriptionId == chat_gpt_endpoint) {
            handleGPTResponse(response);
        } else if (response.subscriptionId == star_wars_endpoint) {
            handleStarWarsResponse(response);
        }
    }
}
```

## Additional Information

- [x] A set of helper interfaces & types were defined.
- [x] The hardhat config was modified to use the `ChainlinkUserConfig` Type.
- [x] Chainlink contracts were installed.
- [x] Compiler version updated to fill chainlink contracts requirements.
- [x] A helper function to get the current network config as a `ChainLinkNetworkUserConfig` type was created.
- [x] A final deploy script to grant all clients access to the gateway was created.
- [x] `standard-version` lib was created for an easy `CHANGELOG` management. Note: Read [this](https://www.conventionalcommits.org/en/v1.0.0/) to properly do commits and auto generate the changelog of your progress.

---
Your ENS/address: [uklok.eth](https://app.ens.domains/uklok.eth)

### CODE REVIEW

**Review by mayanloong — APPROVED** [review]
need fix deployment issue

File: packages/hardhat/contracts/CCExampleClient.sol [hunk 1]
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+import {ICCGatewayClient} from "./interfaces/ICCGatewayClient.sol";
+import {ICCGateway} from "./interfaces/ICCGateway.sol";
+
+contract CCExampleClient is ICCGatewayClient {
```
Comment (mayanloong) [inline packages/hardhat/contracts/CCExampleClient.sol:7]: CCExampleClient didn't call registerRequest()? request() method will fail?

Comment (uklok-agent) [reply packages/hardhat/contracts/CCExampleClient.sol:7]: The gateway is intended to register the available requests outside the smart-contracts so the logic could be abstracted from them.

I've created another PR which is a gateway-toolkit [tasks, scripts, helpers, ...] to streamline all this process as easier as possible for anybody.

**Comment by uklok-agent** [comment]
@allanma88 I checked and somehow there is a conflict with deploys on the `release` standard-version commit. Everything is fine. You can safely skip this warning.

On the other hand. I created the #4 PR and it's an extension of this one and that one have a successfully deployed final commit.

**Pull request merged**